src/sys/unix/uds/mod.rs

```diff
@@ -1,164 +1,167 @@
-        //
-        // `0` is a valid value for `sockaddr_un::sun_family`; it is
-        // `libc::AF_UNSPEC`.
-        //
-        // `[0; 108]` is a valid value for `sockaddr_un::sun_path`; it begins an
-        // abstract path.
-        let mut sockaddr = unsafe { sockaddr.assume_init() };
-
-        sockaddr.sun_family = libc::AF_UNIX as libc::sa_family_t;
-
-        match (bytes.first(), bytes.len().cmp(&sockaddr.sun_path.len())) {
-            // Abstract paths don't need a null terminator
-            (Some(&0), Ordering::Greater) => {
-                return Err(io::Error::new(
-                    io::ErrorKind::InvalidInput,
-                    "path must be no longer than libc::sockaddr_un.sun_path",
-                ));
+/// Converts a Rust `SocketAddr` into the system representation.
+fn unix_addr(address: &SocketAddr) -> (libc::sockaddr_un, libc::socklen_t) {
+    let sockaddr = mem::MaybeUninit::<libc::sockaddr_un>::zeroed();
+
+    // This is safe to assume because a `libc::sockaddr_un` filled with `0`
+    // bytes is properly initialized.
+    //
+    // `0` is a valid value for `sockaddr_un::sun_family`; it is
+    // `libc::AF_UNSPEC`.
+    //
+    // `[0; 108]` is a valid value for `sockaddr_un::sun_path`; it begins an
+    // abstract path.
+    let mut sockaddr = unsafe { sockaddr.assume_init() };
+
+    sockaddr.sun_family = libc::AF_UNIX as libc::sa_family_t;
+
+    #[allow(unused_mut)] // Only used with abstract namespaces.
+    let mut offset = 0;
+    let addr = match address.as_pathname() {
+        Some(path) => path.as_os_str().as_bytes(),
+        #[cfg(target_os = "linux")]
+        None => match address.as_abstract_name() {
+            Some(name) => {
+                offset += 1;
+                name
             }
-            (_, Ordering::Greater) | (_, Ordering::Equal) => {
-                return Err(io::Error::new(
-                    io::ErrorKind::InvalidInput,
-                    "path must be shorter than libc::sockaddr_un.sun_path",
-                ));
-            }
-            _ => {}
-        }
+            None => UNNAMED_ADDRESS,
+        },
+        #[cfg(not(target_os = "linux"))]
+        None => UNNAMED_ADDRESS,
+    };
+
+    // SAFETY: `addr` and `sockaddr.sun_path` are not overlapping and both point
+    // to valid memory.
+    // SAFETY: since `addr` is a valid Unix address, it must not be larger than
+    // `SUN_LEN` bytes, thus we won't overwrite the size of sockaddr.sun_path.
+    // SAFETY: null byte is already written because we zeroed the address above.
+    debug_assert!(addr.len() <= sockaddr.sun_path.len());
+    unsafe {
+        ptr::copy_nonoverlapping(
+            addr.as_ptr(),
+            sockaddr.sun_path.as_mut_ptr().add(offset).cast(),
+            addr.len(),
+        )
+    };
+
+    let mut addrlen = path_offset(&sockaddr) + addr.len();
+    match addr.first() {
+        Some(&0) | None => {}
+        Some(_) => addrlen += 1,
+    }
```

Comment:
I don't understand what's going on here. Would you not add one in the case where it starts with zero?

Reply:
This is based on https://github.com/tokio-rs/mio/blob/2d7b45420b07c49770fc5e8f0fdf8f22bbba1d3a/src/sys/unix/uds/mod.rs#L64-L69.

Reply:
Added a comment.